src/frontend/src/optimizer/plan_node/stream_cdc_table_scan.rs

```diff
@@ -0,0 +1,309 @@
+                distribution: self.distribution(),
+                input_schema: self.base.schema(),
+            });
+            vec.push(("dist", dist));
+        }
+
+        childless_record("StreamCdcTableScan", vec)
+    }
+}
+
+impl StreamNode for StreamCdcTableScan {
+    fn to_stream_prost_body(&self, _state: &mut BuildFragmentGraphState) -> PbNodeBody {
+        unreachable!("stream scan cannot be converted into a prost body -- call `adhoc_to_stream_prost` instead.")
+    }
+}
+
+impl StreamCdcTableScan {
+    pub fn adhoc_to_stream_prost(&self, state: &mut BuildFragmentGraphState) -> PbStreamNode {
+        use risingwave_pb::stream_plan::*;
+
+        let stream_key = self
+            .stream_key()
+            .unwrap_or_else(|| {
+                panic!(
+                    "should always have a stream key in the stream plan but not, sub plan: {}",
+                    PlanRef::from(self.clone()).explain_to_string()
+                )
+            })
+            .iter()
+            .map(|x| *x as u32)
+            .collect_vec();
+
+        // The required columns from the table (both scan and upstream).
+        let upstream_column_ids = self
+            .core
+            .output_and_pk_column_ids()
+            .iter()
+            .map(ColumnId::get_id)
+            .collect_vec();
+
+        // The schema of the snapshot read stream
+        let snapshot_schema = upstream_column_ids
+            .iter()
+            .map(|&id| {
+                let col = self
+                    .core
+                    .get_table_columns()
+                    .iter()
+                    .find(|c| c.column_id.get_id() == id)
+                    .unwrap();
+                Field::from(col).to_prost()
+            })
+            .collect_vec();
+
+        // The schema of the shared cdc source upstream is different from snapshot,
+        // refer to `debezium_cdc_source_schema()` for details.
+        let upstream_schema = {
+            let mut columns = debezium_cdc_source_schema();
+            columns.push(ColumnCatalog::row_id_column());
+            columns
+                .into_iter()
+                .map(|c| Field::from(c.column_desc).to_prost())
+                .collect_vec()
+        };
+
+        let output_indices = self
+            .core
+            .output_column_ids()
+            .iter()
+            .map(|i| {
+                upstream_column_ids
+                    .iter()
+                    .position(|&x| x == i.get_id())
+                    .unwrap() as u32
+            })
+            .collect_vec();
+
+        let batch_plan_node = BatchPlanNode {
+            table_desc: None,
+            column_ids: upstream_column_ids.clone(),
+        };
+
+        let catalog = self
+            .build_backfill_state_catalog(state)
+            .to_internal_table_prost();
+
+        let node_body =
+            // don't need batch plan for cdc source
+            PbNodeBody::StreamScan(StreamScanNode {
```

Comment:
I tend to use a new fragment / proto message for `StreamCdcScanNode`. Of course this can be done in the future. cc. @StrikeW

Reply:
Additionally, we don't need to follow the bad design of `StreamScanNode`, which has 2 inputs - an empty `MergeNode` as placeholder and a strange `BatchScanNode` but not really a `BatchScan`

As I mentioned in https://github.com/risingwavelabs/risingwave/pull/13242#issue-1975861888: This part is shitty but changing it (`StreamScanNode`) will break compatibility. But here we have the chance to change.

Reply:
https://github.com/risingwavelabs/risingwave/issues/13146#issuecomment-1805229754

Added a checklist to track the changes.